### PR TITLE
Register new users using events

### PR DESCRIPTION
**Pull request by camjackson**
~This isn't ready yet, and there are changes happening in other repos too.~

~Just trying to put this out in the open for people to see progress.~

 #112

### CODE REVIEW

**Comment by camjackson** [comment]
Ready for review!

**Comment by camjackson** [comment]
Someone needs to look more closely at the integration tests that have been `xdescribe`'d and `xignore`'d in this PR.

**Review by pameck — APPROVED** [review]
Hey @camjackson, nice work!

These are my observations! I don't see them as a reason not to merge the code, I think we should merge before the branch diverges too much.

Didn't know of `Sinon.sandbox`, so that was quite useful!

### Observations:

* I think it is risky having the tests expect a 500, I'd `.skip` them instead so we know they're pending, but that in case someone enables them again, they still should fail. 

* `MembersController.register` should log the error when failing to publish an event to the stream in case we need to troubleshoot:

  `.catch(() => res.sendStatus(500));` to `.catch(handlerError);`

* I am not sure where the event consumer setup belongs. It seemed a bit odd to have it in the routes file because that's more end-point routing and the consumer is more business-y. 

  My suggestion (which I came up with in 2 mins, so feel free to point all the things I didn't think of :P) :
  
  I'd rather require the consumer in the routes file already setup and would point straight to `MembersService.putMemberInDatabase` instead than `MembersController.putMemberInDatabase` since it is not getting a `request`, `response` object anymore.

**Comment by camjackson** [comment]
- Good point about the `500` tests. Actually the main reason for that was because of not having a local kinesis to connect to, which we've since figured out. So I'll look at fixing those.
 - Yep, I'll put some logging in
 - Yeah I didn't really know where to put that stuff, so I figured I'd copy the existing structure until we've really started to pull the core apart, hoping the right structure would magically appear. I think you're right though, it makes more sense to just bind the service directly to the consumer 👍 I might leave the consumer initialisation in the routes file for now though - there are some changes coming to the API which might have implications for where that initialisation needs to go.

**Comment by camjackson** [comment]
@pameck Oh and yeah, sinon sandbox is great. So much better than having to keep track of all the spies and stubs and clean them up manually!

File: backend/package.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
   "main": "src/app.js",
   "scripts": {
     "ci": "npm run lint && npm test",
+    "kinesis": "node bin/kinesis.js",
```
Comment (yearofthedan) [inline backend/package.json:11]: I can't see where this task is being called from. Is this prep for more to come? If it supports local dev / test can we include an entry in the readme? Possibly rename to local-kinesis.

File: backend/spec/unit/controllers/membersControllerSpec.js [hunk 2]
```diff
@@ -7,115 +7,96 @@ const messagingService = require('../../../src/services/messagingService');
 const membersController = require('../../../src/controllers/membersController');
 const inputValidator = require('../../../src/lib/inputValidator');
 const csvGenerator = require('../../../src/lib/csvGenerator');
+const streamClient = require('../../../src/streamClient');
```
Comment (yearofthedan) [inline backend/spec/unit/controllers/membersControllerSpec.js:10]: No action, just want to complain that these lengthy imports paths make me wish that the tests were colocated with the src.

File: backend/src/config.js [hunk 3]
```diff
@@ -18,12 +18,21 @@ const config = {
...
+  },
+  eventAuthToken: process.env.EVENT_AUTH_TOKEN || 'secret',
 };
 
 if (process.env.NODE_ENV === 'production') {
   config.session.proxy = true;
   config.session.secureCookie = true;
 
+  // Let the SDK derive the endpoint from the region
+  config.aws.kinesisEndpoint = undefined;
```
Comment (yearofthedan) [inline backend/src/config.js]: I'm not sure I like this pattern of sneaky const override. I think I'd rename to base config and then return something like a spread of base + env specific config.

File: backend/src/streamClient.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+const createClient = require('rabblerouser-stream-client');
+const config = require('./config');
+
+const streamClientSettings = {
+  stream: 'rabblerouser_stream',
```
Comment (yearofthedan) [inline backend/src/streamClient.js]: Could these props also be set from config?

File: frontend/src/signup/sagas.js [hunk 5]
```diff
@@ -28,6 +28,7 @@ export function* registerMember({ payload }) {
     yield put(registerSuccess());
     yield call(onSuccess);
   } catch (error) {
+    console.log('ERROR', error);
```
Comment (yearofthedan) [inline frontend/src/signup/sagas.js]: Do we want to console log on the front end?

**Comment by yearofthedan** [comment]
Put my comments in. Looks great and easy to follow. Just a couple of minor questions really.

File: backend/package.json (hunk 1 above)
Comment (camjackson) [reply backend/package.json:11]: Yeah at the moment it's just a thing you would run in a separate terminal, just like you have to have postgres running before you do anything. You're right, it should be documented.

Hopefully that will go away very soon, as we get #129 happening.

Comment (camjackson) [reply backend/package.json:11]: Actually now that I think about it, writing docs for all the different things you need to start is going to be a pain, and then it's all going to get deleted as soon the other PR (docker stuff) happens. So I'm inclined to leave this for now, on the condition that dockerising the other bits a pieces happens straight away.

File: backend/spec/unit/controllers/membersControllerSpec.js (hunk 2 above)
Comment (camjackson) [reply backend/spec/unit/controllers/membersControllerSpec.js:10]: +1

File: backend/src/config.js (hunk 3 above)
Comment (camjackson) [reply backend/src/config.js]: Good idea.

File: backend/src/streamClient.js (hunk 4 above)
Comment (camjackson) [reply backend/src/streamClient.js]: Yep

File: frontend/src/signup/sagas.js (hunk 5 above)
Comment (camjackson) [reply frontend/src/signup/sagas.js]: Oops, that was there for debugging. Should have been taken out.

File: backend/package.json (hunk 1 above)
Comment (yearofthedan) [reply backend/package.json:11]: Thought that might be the case. I'm happy to leave it till #129

**Pull request merged**